src/core/manager-json.c

```diff
@@ -0,0 +1,2264 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <sys/mount.h>
+#include <sys/prctl.h>
+#include <sys/syslog.h>
+
+#include "automount.h"
+#include "build.h"
+#include "confidential-virt.h"
+#include "device.h"
+#include "exec-credential.h"
+#include "fileio.h"
+#include "json.h"
+#include "dbus-unit.h"
+#include "manager-json.h"
+#include "in-addr-prefix-util.h"
+#include "missing_ioprio.h"
+#include "parse-util.h"
+#include "rlimit-util.h"
+#include "process-util.h"
+#include "mount.h"
+#include "path.h"
+#include "scope.h"
+#include "service.h"
+#include "signal-util.h"
+#include "swap.h"
+#include "syslog-util.h"
+#include "timer.h"
+#include "version.h"
+#include "virt.h"
+#include "watchdog.h"
+
+#define JSON_BUILD_DUAL_TIMESTAMP(name, t) \
+        JSON_BUILD_PAIR_FINITE_USEC(name, t.realtime), \
+        JSON_BUILD_PAIR_FINITE_USEC(name "Monotonic", t.monotonic)
+
+static int rlimit_build_json(JsonVariant **ret, const char *name, void *userdata) {
+        const char *is_soft;
+        struct rlimit *rl = userdata;
+        rlim_t x;
+
+        assert(name);
+        assert(ret);
+
+        is_soft = endswith(name, "Soft");
+
+        if (rl)
+                x = is_soft ? rl->rlim_cur : rl->rlim_max;
+        else {
+                struct rlimit buf = {};
+                const char *s, *p;
+                int z;
+
+                /* Chop off "Soft" suffix */
+                s = is_soft ? strndupa_safe(name, is_soft - name) : name;
+
+                /* Skip over any prefix, such as "Default" */
+                assert_se(p = strstrafter(s, "Limit"));
+
+                z = rlimit_from_string(p);
+                assert(z >= 0);
+
+                (void) getrlimit(z, &buf);
+                x = is_soft ? buf.rlim_cur : buf.rlim_max;
+        }
+
+        /* rlim_t might have different sizes, let's map RLIMIT_INFINITY to UINT64_MAX, so that it is the same
+         * on all archs */
+        return json_variant_new_unsigned(ret, x == RLIM_INFINITY ? UINT64_MAX : (uint64_t) x);
```

Comment:
similar here, I think rlimits should be serialize as object with tw fields "hard" and "soft"